// @drenso/homey-zigbee-library/lib/clusters/DoorLockCluster.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GENERAL_RESPONSE_STATUS_ENUM = exports.SET_PIN_CODE_RESPONSE_STATUS_ENUM = exports.USER_TYPE_ENUM = exports.USER_STATUS_ENUM = exports.LOCK_TYPE_ENUM = exports.LOCK_STATE_ENUM = void 0;
const zigbee_clusters_1 = require("zigbee-clusters");
exports.LOCK_STATE_ENUM = {
    not_fully_locked: 0x00,
    locked: 0x01,
    unlocked: 0x02,
    undefined: 0xFF,
};
exports.LOCK_TYPE_ENUM = {
    dead_bolt: 0x00,
    magnetic: 0x01,
    other: 0x02,
    mortise: 0x03,
    rim: 0x04,
    latch_bolt: 0x05,
    cylindrical_lock: 0x06,
    tubular_lock: 0x07,
    interconnected_lock: 0x08,
    dead_latch: 0x09,
    door_furniture: 0x0A,
};
exports.USER_STATUS_ENUM = {
    available: 0x00,
    enabled: 0x01,
    disabled: 0x03,
    not_supported: 0xFF,
};
exports.USER_TYPE_ENUM = {
    unrestricted: 0,
    yearDaySchedule: 1,
    weekdaySchedule: 2,
    master: 3,
    nonAccess: 4,
};
exports.SET_PIN_CODE_RESPONSE_STATUS_ENUM = {
    success: 0,
    general_failure: 1,
    memory_full: 2,
    duplicate_code_error: 3,
};
exports.GENERAL_RESPONSE_STATUS_ENUM = {
    pass: 0,
    fail: 1,
};
const Attributes = {
    lockState: {
        id: 0x0000,
        type: zigbee_clusters_1.ZCLDataTypes.enum8(exports.LOCK_STATE_ENUM),
    },
    lockType: {
        id: 0x0001,
        type: zigbee_clusters_1.ZCLDataTypes.enum8(exports.LOCK_TYPE_ENUM),
    },
    actuatorEnabled: {
        id: 0x0002,
        type: zigbee_clusters_1.ZCLDataTypes.bool,
    },
    numberOfHolidaySchedulesSupported: {
        id: 0x0016,
        type: zigbee_clusters_1.ZCLDataTypes.uint8,
    },
    maxPinCodeLength: {
        id: 0x0017,
        type: zigbee_clusters_1.ZCLDataTypes.uint8,
    },
    minPinCodeLength: {
        id: 0x0018,
        type: zigbee_clusters_1.ZCLDataTypes.uint8,
    },
    enableLogging: {
        id: 0x0020,
        type: zigbee_clusters_1.ZCLDataTypes.bool,
    },
    language: {
        id: 0x0021,
        type: zigbee_clusters_1.ZCLDataTypes.string,
    },
    autoRelockTime: {
        id: 0x0023,
        type: zigbee_clusters_1.ZCLDataTypes.uint32,
    },
    soundVolume: {
        id: 0x0024,
        type: zigbee_clusters_1.ZCLDataTypes.uint8,
    },
};
const CommandsReceived = {
    lockDoor: {
        id: 0x00,
        direction: zigbee_clusters_1.Cluster.DIRECTION_CLIENT_TO_SERVER,
    },
    unlockDoor: {
        id: 0x01,
        direction: zigbee_clusters_1.Cluster.DIRECTION_CLIENT_TO_SERVER,
    },
    toggle: {
        id: 0x02,
        direction: zigbee_clusters_1.Cluster.DIRECTION_CLIENT_TO_SERVER,
    },
    unlockWithTimeout: {
        id: 0x03,
        direction: zigbee_clusters_1.Cluster.DIRECTION_CLIENT_TO_SERVER,
        // args
    },
    getLogRecord: {
        id: 0x04,
        direction: zigbee_clusters_1.Cluster.DIRECTION_CLIENT_TO_SERVER,
        // args
    },
    setPinCode: {
        id: 0x05,
        direction: zigbee_clusters_1.Cluster.DIRECTION_CLIENT_TO_SERVER,
        args: {
            userId: zigbee_clusters_1.ZCLDataTypes.uint16,
            userStatus: zigbee_clusters_1.ZCLDataTypes.enum8(exports.USER_STATUS_ENUM),
            userType: zigbee_clusters_1.ZCLDataTypes.enum8(exports.USER_TYPE_ENUM),
            pin: zigbee_clusters_1.ZCLDataTypes.octstr,
        },
    },
    getPinCode: {
        id: 0x06,
        direction: zigbee_clusters_1.Cluster.DIRECTION_CLIENT_TO_SERVER,
        args: {
            userId: zigbee_clusters_1.ZCLDataTypes.uint16,
        },
    },
    clearPinCode: {
        id: 0x07,
        direction: zigbee_clusters_1.Cluster.DIRECTION_CLIENT_TO_SERVER,
        args: {
            userId: zigbee_clusters_1.ZCLDataTypes.uint16,
        },
    },
    clearAllPinCodes: {
        id: 0x08,
        direction: zigbee_clusters_1.Cluster.DIRECTION_CLIENT_TO_SERVER,
        // args
    },
    setUserStatus: {
        id: 0x09,
        direction: zigbee_clusters_1.Cluster.DIRECTION_CLIENT_TO_SERVER,
        args: {
            userId: zigbee_clusters_1.ZCLDataTypes.uint16,
            userStatus: zigbee_clusters_1.ZCLDataTypes.enum8(exports.USER_STATUS_ENUM),
        },
    },
    getUserStatus: {
        id: 0x0A,
        direction: zigbee_clusters_1.Cluster.DIRECTION_CLIENT_TO_SERVER,
        args: {
            userId: zigbee_clusters_1.ZCLDataTypes.uint16,
        },
    },
    setWeekdaySchedule: {
        id: 0x0B,
        direction: zigbee_clusters_1.Cluster.DIRECTION_CLIENT_TO_SERVER,
        // args
    },
    getWeekdaySchedule: {
        id: 0x0C,
        direction: zigbee_clusters_1.Cluster.DIRECTION_CLIENT_TO_SERVER,
        // args
    },
    clearWeekdaySchedule: {
        id: 0x0D,
        direction: zigbee_clusters_1.Cluster.DIRECTION_CLIENT_TO_SERVER,
        // args
    },
    setYearDaySchedule: {
        id: 0x0E,
        direction: zigbee_clusters_1.Cluster.DIRECTION_CLIENT_TO_SERVER,
        // args
    },
    getYearDaySchedule: {
        id: 0x0F,
        direction: zigbee_clusters_1.Cluster.DIRECTION_CLIENT_TO_SERVER,
        // args
    },
    clearYearDaySchedule: {
        id: 0x10,
        direction: zigbee_clusters_1.Cluster.DIRECTION_CLIENT_TO_SERVER,
        // args
    },
    setUserType: {
        id: 0x14,
        direction: zigbee_clusters_1.Cluster.DIRECTION_CLIENT_TO_SERVER,
        // args
    },
    getUserType: {
        id: 0x15,
        direction: zigbee_clusters_1.Cluster.DIRECTION_CLIENT_TO_SERVER,
        // args
    },
};
const CommandsGenerated = {
    lockDoorResponse: {
        id: 0x00,
        direction: zigbee_clusters_1.Cluster.DIRECTION_SERVER_TO_CLIENT,
        // args
    },
    unlockDoorResponse: {
        id: 0x01,
        direction: zigbee_clusters_1.Cluster.DIRECTION_SERVER_TO_CLIENT,
        // args
    },
    toggleResponse: {
        id: 0x02,
        direction: zigbee_clusters_1.Cluster.DIRECTION_SERVER_TO_CLIENT,
        // args
    },
    unlockWithTimeoutResponse: {
        id: 0x03,
        direction: zigbee_clusters_1.Cluster.DIRECTION_SERVER_TO_CLIENT,
        // args
    },
    getLogRecordResponse: {
        id: 0x04,
        direction: zigbee_clusters_1.Cluster.DIRECTION_SERVER_TO_CLIENT,
        // args
    },
    setPinCodeResponse: {
        id: 0x05,
        direction: zigbee_clusters_1.Cluster.DIRECTION_SERVER_TO_CLIENT,
        args: {
            status: zigbee_clusters_1.ZCLDataTypes.enum8(exports.SET_PIN_CODE_RESPONSE_STATUS_ENUM),
        },
    },
    getPinCodeResponse: {
        id: 0x06,
        direction: zigbee_clusters_1.Cluster.DIRECTION_SERVER_TO_CLIENT,
        args: {
            userId: zigbee_clusters_1.ZCLDataTypes.uint16,
            userStatus: zigbee_clusters_1.ZCLDataTypes.enum8(exports.USER_STATUS_ENUM),
            userType: zigbee_clusters_1.ZCLDataTypes.enum8(exports.USER_TYPE_ENUM),
            pin: zigbee_clusters_1.ZCLDataTypes.octstr,
        },
    },
    clearPinCodeResponse: {
        id: 0x07,
        direction: zigbee_clusters_1.Cluster.DIRECTION_SERVER_TO_CLIENT,
        args: {
            status: zigbee_clusters_1.ZCLDataTypes.enum8(exports.GENERAL_RESPONSE_STATUS_ENUM),
        },
    },
    clearAllPinCodesResponse: {
        id: 0x08,
        direction: zigbee_clusters_1.Cluster.DIRECTION_SERVER_TO_CLIENT,
        // args
    },
    setUserStatusResponse: {
        id: 0x09,
        direction: zigbee_clusters_1.Cluster.DIRECTION_SERVER_TO_CLIENT,
        args: {
            status: zigbee_clusters_1.ZCLDataTypes.enum8(exports.GENERAL_RESPONSE_STATUS_ENUM),
        },
    },
    getUsersStatusResponse: {
        id: 0x0A,
        direction: zigbee_clusters_1.Cluster.DIRECTION_SERVER_TO_CLIENT,
        args: {
            userId: zigbee_clusters_1.ZCLDataTypes.uint16,
            userStatus: zigbee_clusters_1.ZCLDataTypes.enum8(exports.USER_STATUS_ENUM),
        },
    },
    setWeekdayScheduleResponse: {
        id: 0x0B,
        direction: zigbee_clusters_1.Cluster.DIRECTION_SERVER_TO_CLIENT,
        // args
    },
    getWeekdayScheduleResponse: {
        id: 0x0C,
        direction: zigbee_clusters_1.Cluster.DIRECTION_SERVER_TO_CLIENT,
        // args
    },
    clearWeekdayScheduleResponse: {
        id: 0x0D,
        direction: zigbee_clusters_1.Cluster.DIRECTION_SERVER_TO_CLIENT,
        // args
    },
    setYearDayScheduleResponse: {
        id: 0x0E,
        direction: zigbee_clusters_1.Cluster.DIRECTION_SERVER_TO_CLIENT,
        // args
    },
    getYearDayScheduleResponse: {
        id: 0x0F,
        direction: zigbee_clusters_1.Cluster.DIRECTION_SERVER_TO_CLIENT,
        // args
    },
    clearYearDayScheduleResponse: {
        id: 0x10,
        direction: zigbee_clusters_1.Cluster.DIRECTION_SERVER_TO_CLIENT,
        // args
    },
    setUserTypeResponse: {
        id: 0x14,
        direction: zigbee_clusters_1.Cluster.DIRECTION_SERVER_TO_CLIENT,
        // args
    },
    getUserTypeResponse: {
        id: 0x15,
        direction: zigbee_clusters_1.Cluster.DIRECTION_SERVER_TO_CLIENT,
        // args
    },
    operatingEventNotification: {
        id: 0x20,
        direction: zigbee_clusters_1.Cluster.DIRECTION_SERVER_TO_CLIENT,
        // args
    },
    programmingEventNotification: {
        id: 0x21,
        direction: zigbee_clusters_1.Cluster.DIRECTION_SERVER_TO_CLIENT,
        // args
    },
};
class DoorLockCluster extends zigbee_clusters_1.DoorLockCluster {
    static get ATTRIBUTES() {
        return {
            ...super.ATTRIBUTES,
            ...Attributes,
        };
    }
    static get COMMANDS() {
        return {
            ...super.COMMANDS,
            ...CommandsReceived,
            ...CommandsGenerated,
        };
    }
}
exports.default = DoorLockCluster;
