// @drenso/homey-zigbee-library/lib/clusters/ExtendedElectricalMeasurementCluster.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExtendedElectricalMeasurementCluster = void 0;
const zigbee_clusters_1 = require("zigbee-clusters");
class ExtendedElectricalMeasurementCluster extends zigbee_clusters_1.ElectricalMeasurementCluster {
    static get ATTRIBUTES() {
        return { ...super.ATTRIBUTES, ...{
                rmsVoltagePhB: { id: 2309, type: zigbee_clusters_1.ZCLDataTypes.uint16 },
                rmsCurrentPhB: { id: 2312, type: zigbee_clusters_1.ZCLDataTypes.uint16 },
                activePowerPhB: { id: 2315, type: zigbee_clusters_1.ZCLDataTypes.int16 },
                rmsVoltagePhC: { id: 2565, type: zigbee_clusters_1.ZCLDataTypes.uint16 },
                rmsCurrentPhC: { id: 2568, type: zigbee_clusters_1.ZCLDataTypes.uint16 },
                activePowerPhC: { id: 2571, type: zigbee_clusters_1.ZCLDataTypes.int16 },
            } };
    }
}
exports.ExtendedElectricalMeasurementCluster = ExtendedElectricalMeasurementCluster;
zigbee_clusters_1.Cluster.addCluster(ExtendedElectricalMeasurementCluster);
